cpp/src/arrow/engine/substrait/expression_internal.cc

```diff
@@ -57,45 +57,70 @@ Result<compute::Expression> FromProto(const substrait::Expression& expr) {
+      util::optional<compute::Expression> out;
       if (expr.selection().has_expression()) {
-        ARROW_ASSIGN_OR_RAISE(root_expr, FromProto(expr.selection().expression()));
+        ARROW_ASSIGN_OR_RAISE(out, FromProto(expr.selection().expression()));
       }
 
-      const auto& ref = expr.selection().direct_reference();
-      switch (ref.reference_type_case()) {
-        case substrait::Expression::ReferenceSegment::kStructField: {
-          if (ref.struct_field().has_child()) break;
-
-          FieldRef out(ref.struct_field().field());
+      const auto* ref = &expr.selection().direct_reference();
+      while (ref != NULLPTR) {
+        switch (ref->reference_type_case()) {
+          case substrait::Expression::ReferenceSegment::kStructField: {
+            auto index = ref->struct_field().field();
+            if (!out) {
+              // Root StructField (column selection)
+              out = compute::field_ref(FieldRef(index));
+            } else if (auto out_ref = out->field_ref()) {
+              // Nested StructFields on the root (selection of struct-typed column
+              // combined with selecting struct fields)
+              out = compute::field_ref(FieldRef(*out_ref, index));
+            } else if (out->call() && out->call()->function_name == "struct_field") {
+              // Nested StructFields on top of an arbitrary expression
+              std::static_pointer_cast<arrow::compute::StructFieldOptions>(
+                  out->call()->options)
+                  ->indices.push_back(index);
+            } else {
+              // First StructField on top of an arbitrary expression
+              out = compute::call("struct_field", {std::move(*out)},
+                                  arrow::compute::StructFieldOptions({index}));
+            }
 
-          if (root_expr) {
-            if (auto root_ref = root_expr->field_ref()) {
-              out = FieldRef(std::move(out), *root_ref);
+            // Segment handled, continue with child segment (if any)
+            if (ref->struct_field().has_child()) {
+              ref = &ref->struct_field().child();
             } else {
-              // FIXME add struct_field compute function to handle
-              // field references into expressions
-              break;
+              ref = NULLPTR;
             }
+            break;
           }
-          return compute::field_ref(std::move(out));
-        }
+          case substrait::Expression::ReferenceSegment::kListElement: {
+            if (!out) {
+              // Root ListField (illegal)
+              return Status::Invalid(
+                  "substrait::ListElement cannot take a Relation as an argument");
+            } else {
+              // ListField on top of an arbitrary expression
+              out = compute::call(
+                  "list_element",
+                  {std::move(*out), compute::literal(ref->list_element().offset())});
+            }
```

Comment:
Style guide: no redundant `else` blocks
```suggestion
            }

            // ListField on top of an arbitrary expression
            out = compute::call(
                "list_element",
                {std::move(*out), compute::literal(ref->list_element().offset())});
```

Reply:
Ah, that's fair enough. Wasn't aware of that style guide clause.